tests/posix/common/src/nanosleep.c

```diff
@@ -199,23 +178,30 @@ static void common(const uint32_t s, uint32_t ns)
 	zassert_equal(rem.tv_nsec, 0, "actual: %d expected: %d",
 		rem.tv_nsec, 0);
 
+	uint64_t actual_ns = k_cyc_to_ns_ceil64((now - then));
```

Comment:
I'm still approving, but this is where it's going to break: when we have a system where the cycle clock is slower than the CPU clock then we'll have to add a fudge factor for lower bounds that aren't met.  That'll generally be board-specific; I think one of my reel_board's has exhibited that behavior in the past.

Reply:
Well, "break" is debatable. From [the spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/nanosleep.html):

> The suspension time may be longer than requested because the argument value is rounded up to an integer multiple of the sleep resolution or because of the scheduling of other activity by the system.

Which clearly applies here.

I considered adding a negative value in `rmtp` to reflect the overrun w.r.t. `rqtp` but the Linux & macOS do not do that, and it seems that `rmtp` is only modified when `nanosleep(2)` is interrupted by a signal.

Reply:
> Well, "break" is debatable. From [the spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/nanosleep.html):
> 
> > The suspension time may be longer than requested because the argument value is rounded up to an integer multiple of the sleep resolution or because of the scheduling of other activity by the system.

Let me try again:

Assume the CPU clock that's used for the nanosleep busy-wait runs at 10 MHz and is precise: there are 10^7 ticks in one second.

There is no access to this as a clock, so a different clock is used for the system clock that counts cycles.  Assume it runs at 10 kHz and has a -500 ppm error: it ticks 9995 times in one second.

Then the duration measured for a 1 s sleep taking 10^7 CPU cycles will be 999500 us, which will appear to violate the spec because it's shorter than expected.  The test will fail, even though the code does exactly what it's supposed to do.  That's what I mean by "break".

This mostly shows up in Nordic because other platforms have synchronous CPU and cycle clocks, but in theory it can apply to any platform.  The frequency tolerance for Nordic LFRC is 5%, and for HFRC is 8% so a 13% error between the CPU and cycle clocks can be observed.  It's pretty likely at least one board in some test environment somewhere will display a negative ppm error and fail the test.

Reply:
> This mostly shows up in Nordic because other platforms have synchronous CPU and cycle clocks, but in theory it can apply to any platform. The frequency tolerance for Nordic LFRC is 5%, and for HFRC is 8% so a 13% error between the CPU and cycle clocks can be observed. It's pretty likely at least one board in some test environment somewhere will display a negative ppm error and fail the test.

Thanks for clarifying. I think that is a definite concern when using one clock domain to try and synchronize a separate clock domain. That is not what is happening here. In the implementation, I took out the busy wait and opted to use `k_usleep()` instead, which synchronises in the same (albeit slow as heck) clock domain.

Assumption 1: `k_usleep()` does not use the cycle counter that `k_busy_wait()` uses.

Assumption 1 seems to be valid but please correct me if I'm wrong.

Assumption 2: calling `k_usleep()` for any non-zero duration would incur at least 1 expiration of the cycle clock

My assumption was, that by doing so, there would be a guarantee that the cycle clock must expire at least once (it is likely more than that).

It would appear that Assumption 2 also holds.

So I am skeptical that (now - then) will ever be a negative value in this test.

I realize that nanosleep is incredibly inaccurate this way, but it is both compliant with the spec and the tests do not break.

Given that Zephyr is not relying on nanosleep anywhere yet, I feel that the inaccuracy is a lesser evil than requiring all platforms to try and tune their clock-tolerances.

I think the issue here, is that Nordic put a lot of work into trying to tune nanosleep based on clock tolerances, but it was difficult to get tests to pass consistently. Tests also do not pass consistently in qemu as well, or on other platforms I was using. Using clock tolerances is a bit of a slippery slope.

The posixly-correct way to fix this would be to provide access to a collection of clocks and their resolutions via `clock_getres()`, then call `clock_nanosleep()` using the clock with the best resolution.

Since we don't (yet) have `clock_getres()`, and I believe we do not have the ability to use an arbitrary clock to un-suspend a thread, and also since we do not have a calibrated per-cpu delay-loop, I'm ok with using `k_usleep()` until other more reliable and more accurate options are available.

I realize it's a lot to ask, but can you be ok with that as well for the next release?

Reply:
I'd missed that the implementation had also changed away from busy-wait.  I haven't seen anything here that makes me remove my existing approval, but we need @pfalcon to approve this.

For the other stuff, I just want to point to #19030 and its parent #19282.  Clocks and time are very important to me.